### PR TITLE
Multi download

### PR DESCRIPTION
**Pull request by westnordost**

### CODE REVIEW

**Comment by westnordost** [comment]
Tried download in my neighbourhood. **BUT** with my new Samsung S10e which has a powerful processor etc.

Type | Time
------------ | -------------
Download of 47450 nodes, 7586 ways and 238 relations|10s
Analyzing data set (all in below table)|5s
Persisting created quests|3s

I marked in fat which are (still) quite slow:

Quest type | Time
------------ | -------------
**AddPlaceName** | **532ms** - extensive filter which applies to nodes, ways and relations + using the feature dictionary for every candidate. There are performance improvements that could be made on the feature dictionary
AddOneway | 86ms
AddBusStopName | 54ms
AddBusStopRef | Skipped because it is disabled for this country
AddIsBuildingUnderground | 8ms
**AddHousenumber** | **401ms** - extensive geometry checks. I am surprised it is this fast
MarkCompletedHighwayConstruction | 9ms
AddReligionToPlaceOfWorship | 33ms
AddParkingAccess | 26ms
AddRecyclingType | 24ms
AddSport | 28ms
AddRoadSurface | 293ms
AddMaxHeight | 85ms
AddRailwayCrossingBarrier | 41ms
AddPostboxCollectionTimes | 32ms
**AddOpeningHours** | **1657ms** - extensive filter which applies to nodes, ways and relations + using the feature dictionary for every candidate + using the opening hours parser for every candidate
DetailRoadSurface | 26ms
AddBikeParkingCapacity | 39ms
AddOrchardProduce | 7ms
AddBuildingType | 123ms
AddProhibitedForPedestrians | 216ms
AddCrossingType | 36ms
AddCrossingIsland | 58ms
AddBuildingLevels | 35ms
AddBusStopShelter | 42ms
AddParkingFee | 29ms
AddMotorcycleParkingCapacity | 23ms
AddPathSurface | 97ms
AddTracktype | 11ms
AddMaxWeight | 41ms
AddBikeParkingType | 22ms
AddStepsRamp | 11ms
AddWheelchairAccessToilets | 22ms
AddPlaygroundAccess | 21ms
AddToiletAvailability | 53ms
AddFerryAccessPedestrian | 5ms
AddFerryAccessMotorVehicle | 5ms
DetermineRecyclingGlass | 18ms
AddWayLit | 111ms
AddToiletsFee | 28ms
AddBikeParkingCover | 23ms
AddTactilePavingCrosswalk | 70ms
AddTrafficSignalsSound | 20ms
AddTrafficSignalsVibration | 23ms
AddRoofShape | 8ms
AddWheelchairAccessPublicTransport | 36ms
AddWheelchairAccessOutside | 30ms
AddTactilePavingBusStop | 49ms
AddBridgeStructure | 5ms
AddReligionToWaysideShrine | 21ms
AddCyclewaySegregation | 12ms
MarkCompletedBuildingConstruction | 6ms
AddGeneralFee | 35ms
AddSelfServiceLaundry | 25ms
AddStepsIncline | 7ms
AddHandrail | 12ms
AddStepCount | 6ms
AddInformationToTourism | 21ms
AddAtmOperator | 18ms
AddChargingStationOperator | 17ms
AddClothingBinOperator | 26ms
AddIsDefibrillatorIndoor | 18ms
AddSummitRegister | 39ms
AddCyclewayPartSurface | 23ms
AddFootwayPartSurface | 15ms
AddMotorcycleParkingCover | 20ms
AddFireHydrantType | 18ms
AddParkingType | 24ms
AddPostboxRef | Skipped because it is disabled for this country
AddBoardType | 22ms
AddPowerPolesMaterial | 19ms
AddCarWashType | 22ms
AddBenchStatusOnBusStop | 37ms
AddBenchBackrest | 18ms
AddTrafficSignalsButton | 19ms

It is fast, but here I had to wait 20 seconds for the first quest to appear on the map. On Wifi and with a good smartphone! So, this is still kind of slow.

I intend to do the following:
- basically every smartphone has at minimum a dual core and some have an octa core processor. So, I should use multithreading to speed up analyzing by factor 2 to factor 8
- by default (auto-sync download much smaller areas than currently) because the download time + analyzing time + persisting time directly correlates to the data size now. Before, there was a certain minimum time you had to wait for overpass to respond anyway. Maybe 1/4 or even 1/16. The results above are from zooming out as far as the app lets me and then downloading within a city center
- convert the rest of the quest types if possible to quest types that are analyzed locally. These will be slow ones, though.

**Comment by smichel17** [comment]
How large of an area is this downloading? Would this approach work?

- Download a quite small area around the user's current location
- Use the number of nodes inside this area to estimate the density in the user's location
- Download a larger area (size based on the estimated density)

Benefits:

1. Small area is processed more quickly, so we can show quests right away
    - Added bonus: they will be very close, so the user doesn't have to walk to solve them
2. Density estimate helps make sure you won't send a request for too many nodes.
3. Two rounds allows some overlap (process first round of data while second round is downloading)

**Comment by westnordost** [comment]
Good idea, but not right away as part of this PR. There is a lot to change after the core of the download changed. The current download strategies, the progress bar, the autosync (when to download) etc.

**Comment by westnordost** [comment]
> How large of an area is this downloading

About 20 km²

**Comment by matkoniecz** [comment]
> It is fast, but here I had to wait 20 seconds for the first quest to appear on the map.

Maybe it would be feasible to show processed quests without waiting for later ones?

**Comment by westnordost** [comment]
Maybe, I'll try that. But the last test I did, the whole of analyzing the data set took 3s (because of multithreading). Showing each quest as it gets downloaded will slow it down considerably because every update to the quest pins being displayed on the map involves replacing all quest pins that are currently displayed with the new quest pins that should be displayed now. (See https://github.com/tangrams/tangram-es/pull/2112 )

**Comment by smichel17** [comment]
How long does a render add? As a compromise, you could render with exponential back-off. So, after 1, 10, 100, 1000, etc quests are found. As an optimization, once the core functionality is done.

**Comment by westnordost** [comment]
TODO after merge: Remove the sentence _"That is, with the &lt;a href=\"https://wiki.openstreetmap.org/wiki/Overpass_API\"&gt;Overpass API&lt;/a&gt; for downloading quests  (which is anonymous) and with the &lt;a href=\"https://wiki.openstreetmap.org/wiki/API_v0.6\"&gt;OSM API&lt;/a&gt; for uploading changes.&lt;br/&gt;"_ from `privacy_html` in `strings.xml` from all languages.

**Comment by HolgerJeromin** [comment]
You do download from the main osm api?
Getting (raw) the data from overpass would reduce the  load on the main servers.

**Comment by matkoniecz** [comment]
Using this services for editing is perfectly fine, it is an editing API after all. Overpass severs may desync, serve stale data, have big replication, be blocked in Russia, serve 429 errors and all kinds if funny stuff.

And in the end in both cases servers operated by OSM community need to serve this request, so this difference is minor.

**Comment by westnordost** [comment]
gettorf

**Comment by westnordost** [comment]
Erm, how do I measure how long it takes to download an URL? If I enter this URL into Firefox, it won't download anything apparently, same with Chrome. (Is this some stupid security feature?)

In `curl`, I didn't see any possibility to output the time spent to download it.

This is Gettorf, which takes 30s to download on that smartphone. **Apparently** it takes about 1s to download on my computer, so the download itself is perhaps not the big time sink but the available RAM or something
https://api.openstreetmap.org/api/0.6/map?bbox=10.107421875,54.2908816465701,10.1513671875,54.3421488644834

**Comment by matkoniecz** [comment]
I get

```
curl  https://api.openstreetmap.org/api/0.6/map?bbox=10.107421875,54.2908816465701,10.1513671875,54.3421488644834
You requested too many nodes (limit is 50000). Either request a smaller area, or use planet.osm
``

**Comment by matkoniecz** [comment]
And at least on linux `curl` with output send somewhere shows timing.

```
curl https://github.com/westnordost/StreetComplete/pull/1901 > /dev/null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  292k    0  292k    0     0   196k      0 --:--:--  0:00:01 --:--:--  196k
```

**Comment by westnordost** [comment]
Some new results:
- There was a performance problem in the osmapi library when parsing map data. Parsing is now about 100% faster (turns out that 90% of download time is actually parsing time). It is still slow though. Apparently, parsing a date is slow in Java, at least with `SimpleDateFormat`. Should do tests whether the `java.time` stuff introduced since Java 8 is faster.
- setting large initial capacities for `MapData` has no measurable effect on performance

**Comment by HolgerJeromin** [comment]
> Apparently, parsing a date is slow in Java, at least with SimpleDateFormat. 

Do you always need the date as a real date object? Perhaps at some points this is sufficient to keep it as a string.

**Comment by westnordost** [comment]
stats

**Comment by matkoniecz** [comment]
> first download a tiny area

This has additional benefit of user not waiting for long time to see anything. 115 seconds is long enough for a new user to consider app as broken.

Looking at graphs: download/parsing and persisting scaling roughly linearly with dataset size and OSM quest count. So initial quick scan may help both with first impression and for finding first nearby targets in a new area. 18200 quests in 115 seconds is nice, but 10 nearby quests in 5 seconds is a much better for start.

**Comment by smichel17** [comment]
Follow-up question: how large is the stored quest data? Does it (or should it) interact with the "map cache size" setting?

**Comment by westnordost** [comment]
It doesn't. Quests are not considered part of the map. Map cache is only for the background map. Unsolved quests are

**Comment by smichel17** [comment]
>  Unsolved quests are

I think you accidentally

**Comment by westnordost** [comment]
... deleted automatically after a certain interval

**Comment by westnordost** [comment]
Hmm, I think I am done.

Anyone want to review the code or give it a go in a test?

I implemented what @smichel17 and @matkoniecz suggested. On my Samsung S10e, download is so fast that after I tapped through the tutorial, all the quests are already there. 5s for initial scan (zoom 16 tile around my location - 360x360m) yielding 140 quests, then another 8 seconds for the download in 600m radius around my location, yielding another 2200 quests.

The download logic works like that: First download the z16 tile which encloses the user's location (around 360x360m), then from that, calculate the quest density. The download strategy when in WiFi targets downloading around 1000 quests around the user, so it calculates the estimated download radius based on the measured quest density and then issues another download. In my case, that was around 600m. The mobile data download strategry aims at having around 500 quests around the user, so downloaded radii will be smaller.

**Comment by westnordost** [comment]
Interesting classes:
* [AVariableRadiusStrategy.kt](https://github.com/westnordost/StreetComplete/blob/6f8cf82e0671a96ddc4e1623f0cf80dec27dea0e/app/src/main/java/de/westnordost/streetcomplete/data/download/AVariableRadiusStrategy.kt)
*

File: app/src/main/java/de/westnordost/streetcomplete/data/download/AVariableRadiusStrategy.kt [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    protected abstract val desiredQuestCountInVicinity: Int
+
+    override fun getDownloadBoundingBox(pos: LatLon): BoundingBox? {
+        val tileZoom = ApplicationConstants.QUEST_TILE_ZOOM
+
+        val thisTile = pos.enclosingTile(tileZoom)
+        val hasMissingQuestsForThisTile = hasMissingQuestsFor(thisTile.toTilesRect())
+
+        // if at the location where we are, there is nothing yet, first download the tiniest
+        // possible bbox (~ 360x360m) so that we can estimate the quest density
```
Comment (smichel17) [inline app/src/main/java/de/westnordost/streetcomplete/data/download/AVariableRadiusStrategy.kt:33]: > 5s for initial scan (zoom 16 tile around my location - 360x360m) yielding 140 quests

~~This is great if the user already has location services turned on when opening the app. However, I keep location services disabled, so I often open SC, tap the gps button, enable them, and then return to the app. 5 seconds is definitely an improvement over the status quo, but… I am having trouble phrasing this without sounding silly ("5 seconds is still a long time to wait") — objectively it is not long to wait but long enough that a significant amount of users will close the app.~~

~~I wonder if it is possible to do an overpass query for an even smaller location and get something like \~10 quests in 1 second.~~

After trying it out, actually getting a gps fix took so much longer than the download, in my area at least, that I don't think this matters very much.

**Review by smichel17 — COMMENTED** [review]
I can't comment on the parsing without spending more time learning OSM details than I want, but I found one potential issue.

File: app/src/main/java/de/westnordost/streetcomplete/data/download/AVariableRadiusStrategy.kt [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        val areaInKm = boundingBox.area()
+        val visibleQuestCount = visibleQuestsSource.getAllVisibleCount(boundingBox)
+        return visibleQuestCount / areaInKm
+    }
+
+    /** return if there are any quests in the given tiles rect that haven't been downloaded yet */
+    private fun hasMissingQuestsFor(tilesRect: TilesRect): Boolean {
+        val questExpirationTime = ApplicationConstants.REFRESH_QUESTS_AFTER
+        val ignoreOlderThan = max(0, System.currentTimeMillis() - questExpirationTime)
+        return !downloadedTilesDao.get(tilesRect, ignoreOlderThan).contains(DownloadedTilesType.QUESTS)
```
[the remark this answers is not in the document]
Comment (westnordost) [reply app/src/main/java/de/westnordost/streetcomplete/data/download/AVariableRadiusStrategy.kt:66]: `downloadedTilesDao` saves in which map tiles (at zoom 16) quests have already been downloaded. When trying to download quests in only tiles that have been recorded to have been downloaded already, download (manual or automatic) will not start, showing the message "No further quests available".

The lines you commented cause that tiles that have been registered to have been downloaded more than `REFRESH_QUESTS_AFTER` time ago (3 days) are not counted as downloaded anymore.

If every quest within the small area has been solved, the following things will happen:
1. if the `downloadedTilesDao` reports that this tile has not already been downloaded (less than 3 days ago), first the tiny area is downloaded again.
2. After that download finished (if any), the app will calculate that the quest density is 0 here. **This will lead to a division by zero error, gonna fix that now, thank you!** Since the quest quest density can then not be determined (is 0), the app will subsequently try to download the maximum allowed area by the download strategy (`maxDownloadAreaInKm2`). F.e. for the Wifi strategy it is 12 km². 
3. Before downloading however, it will check again if all the tiles that enclose the download area have already been downloaded (asking `downloadedTilesDao`). If they have, the message "No further quests available" will be shown.

**Comment by westnordost** [comment]
Thanks for trying it out, @smichel17 !

File: app/src/main/java/de/westnordost/streetcomplete/data/download/AVariableRadiusStrategy.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (smichel17) [reply app/src/main/java/de/westnordost/streetcomplete/data/download/AVariableRadiusStrategy.kt:66]: > Before downloading however, it will check again if all the tiles that enclose the download area have already been downloaded (asking `downloadedTilesDao`).

Got'cha. I didn't check the code outside this file so I wasn't sure if `downloadedTileDao` was keeping track of node download history or only the quests in them.

[the remark this answers is not in the document]
Comment (westnordost) [reply app/src/main/java/de/westnordost/streetcomplete/data/download/AVariableRadiusStrategy.kt:66]: Neither. It just records whether a download was done. :-)

**Comment by westnordost** [comment]
- [AddHousenumber](https://github.com/westnordost/StreetComplete/blob/6f8cf82e0671a96ddc4e1623f0cf80dec27dea0e/app/src/main/java/de/westnordost/streetcomplete/quests/housenumber/AddHousenumber.kt) - complex filter checks programmatically

File: app/src/main/java/de/westnordost/streetcomplete/quests/housenumber/AddHousenumber.kt [hunk 3]
```diff
@@ -38,111 +31,90 @@ class AddHousenumber(private val overpass: OverpassMapDataAndGeometryApi) : OsmE
...
-            if (createdQuests++ >= DEFAULT_MAX_QUESTS) break
-
-            val addrContainedInBuilding = getPositionContainedInBuilding(building.geometry, addrPositions)
-            if (addrContainedInBuilding != null) {
-                addrPositions.remove(addrContainedInBuilding)
-                continue
-            }
+        /** exclude buildings that intersect with the bounding box because it is not possible to
+           ascertain for these if there is an address node within the building - it could be outside
+           the bounding box */
```
Comment (westnordost) [inline app/src/main/java/de/westnordost/streetcomplete/quests/housenumber/AddHousenumber.kt:62]: Unfortunately, the problem here is that in small bounding boxes, quite a few buildings will be excluded because of this. If a housenumber quest was downloaded before as part of a big bounding box and then another small download is being executed, housenumber quests on buildings on the edge will vanish.

The prior solution was to expand the bounding box to include all the buildings and then download the housenumber nodes based on that enlarged bounding box. With the solution now, this is not possible (or: not expedient) anymore - we have that one bounding box with the data and we should create all the quests based on the downloaded data.
